File: homework1/src/warmup.js
const rp = require('request-promise');


const crypto = require('crypto');

exports.change = (amount) => {
  if (amount === 0) {
    return [0, 0, 0, 0];
  }

  if (amount < 0) {
    throw new RangeError('Amount cannot be negative');
  }

  const initAmount = amount;
  let newAmount;

  const quarters = Math.floor(initAmount / 25);
  newAmount = initAmount % 25;

  const dimes = Math.floor(newAmount / 10);
  newAmount %= 10;

  const nickels = Math.floor(newAmount / 5);
  newAmount %= 5;

  const pennies = newAmount;

  return [quarters, dimes, nickels, pennies];
};


exports.stripQuotes = input => input.replace(/['"]/g, '');


exports.scramble = input => input.split('').sort(() => 0.5 - Math.random()).join('');


exports.powers = (base, limit, callback) => {
  for (let i = 1; i <= limit; i *= base) {
    callback(i);
  }
};


exports.powersGenerator = function* g(base, limit) {
  for (let x = 1; x <= limit; x *= base) {
    yield x;
  }
};


exports.say = (word) => {
  if (!word) {
    return '';
  }
  return nextword => (!nextword ? word : exports.say(`${word} ${nextword}`));
};

exports.interleave = (a, ...b) => {
  const result = [];
  for (let i = 0; i <= Math.max(a.length, b.length); i += 1) {
    if (i < a.length) {
      result.push(a[i]);
    }
    if (i < b.length) {
      result.push(b[i]);
    }
  }
  return result;
};

exports.cylinder = (spec) => {
  let { radius = 1, height = 1 } = spec;
  const surfaceArea = () => ((2 * Math.PI * radius * height) + (2 * Math.PI * (radius ** 2)));
  const volume = () => (Math.PI * (radius ** 2) * height);
  const stretch = (grow) => { height *= grow; };
  const widen = (factor) => { radius *= factor; };
  const toString = () => `Cylinder with radius ${radius} and height ${height}`;
  return Object.freeze({
    get radius() { return radius; },
    get height() { return height; },
    surfaceArea,
    volume,
    stretch,
    widen,
    toString,
  });
};

exports.makeCryptoFunctions = (cryptoKey, cryptoAlgorithm) => {
  function encrypt(text) {
    const encode = crypto.createCipher(cryptoAlgorithm, cryptoKey);
    let encryptIt = encode.update(text, 'utf8', 'hex');
    encryptIt += encode.final('hex');
    return encryptIt;
  }
  function decrypt(text) {
    const decode = crypto.createDecipher(cryptoAlgorithm, cryptoKey);
    let decyptIt = decode.update(text, 'hex', 'utf8');
    decyptIt += decode.final('utf8');
    return decyptIt;
  }
  return [encrypt, decrypt];
};

exports.randomName = ({ gender, region }) => rp({
  uri: 'https://uinames.com/api',
  qs: { gender, region },
}).then(person => `${person.surname}, ${person.name}`);

File: homework2/warmup.py
import random
import math
import requests
from cryptography.fernet import Fernet

def change(amount):
    if amount == 0:
        return (0, 0, 0, 0)
    if amount < 0:
        raise ValueError('amount cannot be negative')

    init_amount = amount
    quarters = int(init_amount / 25)
    new_amount = init_amount % 25

    dimes = int(new_amount / 10)
    new_amount %= 10

    nickels = int(new_amount / 5)
    new_amount %= 5

    pennies = int(new_amount)

    return (quarters, dimes, nickels, pennies)


def strip_quotes(strip):
    return strip.replace('"', '').replace("'", '')

def scramble(scramb):
    return ''.join(random.sample(scramb, len(scramb)))

def powers(base, limit):
    num = 1
    while num <= limit:
        yield num
        num *= base
  # for (let x = 1; x <= limit; x *= base) {
  #   yield x;
  # }

def triples(limit):
    return [(a, b, c) for c in range(1, limit+1) for b in range(1, c) for a in range(b) if a**2 + b**2 == c**2]

def say(word=None):
    if word is None:
        return ''
    def say_more(nextword=None):
        return word if nextword is None else say(f'{word} {nextword}')
    return say_more


def interleave(first, *b):
    result = []
    for i in range(max(len(first), len(b))):
        if i < len(first):
            result.append(first[i])
        if i < len(b):
            result.append(b[i])
    return result

class Cylinder():
    def __init__(self, radius=1, height=1):
        self.radius = radius
        self.height = height
    @property
    def volume(self):
        return math.pi * (self.radius ** 2) * self.height
    @property
    def surface_area(self):
        return (2 * math.pi * self.radius * self.height) + (2 * math.pi * (self.radius ** 2))

    def widen(self, factor):
        self.radius *= factor
        return self

    def stretch(self, factor):
        self.height *= factor
        return self

def make_crypto_functions(key):
    f_key = Fernet(key)
    def encrypt(text):
        return f_key.encrypt(text)
    def decrypt(encryption):
        return f_key.decrypt(encryption)

    return (encrypt, decrypt)



def random_name(**kwargs):
    params = {'gender' : kwargs.get('gender'), 'region' : kwargs.get('region'), 'amount': 1}
    response = requests.get("http://uinames.com/api", params=params)
    if response.status_code in range(200, 300):
        person_info = response.json()
        return f'{person_info["surname"]}, {person_info["name"]}'
    raise ValueError(response.text)
